mint-client/src/clients/user.rs

```diff
@@ -251,14 +248,27 @@ impl UserClient {
         address
     }
 
-    pub fn select_and_spend_coins(
+    pub async fn select_and_spend_coins<R: RngCore + CryptoRng>(
         &self,
         amount: Amount,
-    ) -> Result<Coins<SpendableCoin>, MintClientError> {
+        rng: R,
+    ) -> Result<Coins<SpendableCoin>, ClientError> {
```

Comment:
Rather than having `select_and_spend_coins` side-effect by making change through submitting and awaiting for a transaction, it would be better for users if we return `(Coins<SpendableCoin>, Amount)` where `Amount` is the amount of change we should ask for as an output.

We can eliminate the `create_coin_input` function which is a bit dangerous to call and just inline the calls in `peg_out` and `fund_outgoing_ln_contract` allowing them to construct the change outputs from the `Amount` returned by `select_and_spend_coins`.

The idea would be rather than submitting and waiting for 2 transactions, just let the `peg_out` and `fund_outgoing_ln_contract` functions create a change output in 1 transaction.

We may also wish to add a new function for transferring minted coins between users that could wait until we have made a transaction to give change to ourselves (like your code currently does) then the other user must `reissue`.  Maybe there is a way to do this without 2 transactions a well...

Reply:
> The idea would be rather than submitting and waiting for 2 transactions, just let the peg_out and fund_outgoing_ln_contract functions create a change output in 1 transaction.

That's the idea: one transaction creates both a LN/peg-out output and the change.

I feel like what we really want is something like PSBTs for federation transactions. Maybe there could be a function `fund_transaction(&self, tx: Transaction) -> Transaction` that adds the inputs/outputs and marks the inputs as pending used and the outputs as to be fetched. 

That actually raises another question: how to determine that a tx won't go through anymore. Maybe we should add a valid-if-agreed-upon-before-epoch field. Then we could remove the pending states after the timeout.

Reply:
Yeah, a PSBT-like format would be a bit over the top, but being able to construct a balanced tx step by step would be nice. Multi-user tx are out of scope for now.

Reply:
I suppose it would be more user-friendly to not allow them to instantiate a `Transaction` unless the inputs and outputs are equal to prevent them from losing funds.  Methods within the `UserClient` could pass around a `TransactionBuilder` to add inputs and outputs before signing the final transaction.